ingress/loadbalancer.go

```diff
@@ -17,19 +37,181 @@ type LoadBalancerEntry struct {
+	lb.update(LoadBalancerUpdate{Entries: []LoadBalancerEntry{}})
+
+	configLocation := lb.confDir + "/nginx.conf"
+	log.Infof("Using config file %s", configLocation)
+	cmd := exec.Command(lb.binary, "-c", configLocation)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
+
+	lb.cmd = cmd
+	err := cmd.Start()
+	if err != nil {
+		log.Error("Unable to start nginx", err, cmd)
+		return err
+	}
+
+	go func() {
+		log.Info("Nginx has exited: ", cmd.Wait())
+	}()
+
+	log.Infof("Nginx pid %s", cmd.Process.Pid)
+	return nil
+}
+
+func (lb *nginxLoadBalancer) Stop() error {
+	log.Info("Shutting down process %s", lb.cmd.Process.Pid)
+	err := lb.signaller.Sigquit(lb.cmd.Process)
+	if err != nil {
+		log.Error("Error shutting down nginx", err)
+		return err
+	}
+	return nil
+}
+func (lb *nginxLoadBalancer) Update(entries LoadBalancerUpdate) (bool, error) {
+	updated, err := lb.update(entries)
+	if err != nil {
+		return updated, err
+	}
+	if updated {
+		err = lb.signaller.Sighup(lb.cmd.Process)
+		if err != nil {
+			log.Warn("Configuration was wirtten but unable to signal nginx to reload configuration", err)
+			return false, err
+		}
+	}
+	return updated, err
+}
+
+func (lb *nginxLoadBalancer) update(entries LoadBalancerUpdate) (bool, error) {
+	log.Debugf("Updating loadbalancer %s", entries)
+	file, err := lb.createConfig(entries)
+	if err != nil {
+		return false, err
+	}
+	existing, err := ioutil.ReadFile(lb.confDir + "/nginx.conf")
+	if err != nil {
+		log.Info("Unable to read existing nginx config file. Assuming it needs creating for the first time.", err)
+		return writeFile(lb.confDir+"/nginx.conf", file)
+
+	}
+	diffOutput, err := diff(existing, file)
+	if err != nil {
+		log.Warn("Unable to diff nginx files", err)
+		return false, err
+	}
+	log.Infof("Diff output %s", string(diffOutput))
+	if len(diffOutput) == 0 {
+		log.Info("Configuration has not changed")
+		return false, nil
+	}
+
+	log.Info("Configuration is different. Updating configuration.")
+	_, err = writeFile(lb.confDir+"/nginx.conf", file)
+	if err != nil {
+		log.Warn("Unable to write new nginx configuration", err)
+		return false, err
+	}
+	return true, nil
+}
+
+func writeFile(location string, contents []byte) (bool, error) {
+	err := ioutil.WriteFile(location, contents, 0644)
+	if err != nil {
+		log.Warn("Unable to write new nginx config", err)
+		return false, err
+	}
+	return true, nil
+}
+
+func (lb *nginxLoadBalancer) createConfig(update LoadBalancerUpdate) ([]byte, error) {
+	tmpl, err := template.New("nginx.tmpl").ParseFiles("./nginx.tmpl")
+	if err != nil {
+		log.Warn("Unable to generate nginx configurtion. It will be out of date.", err)
+		return []byte{}, err
+	}
+
+	var output bytes.Buffer
+	validEntries := FilterInvalidEntries(update.Entries)
+	err = tmpl.Execute(&output, loadBalancerTemplate{Port: lb.port, Entries: validEntries, WorkingDir: lb.confDir})
+
+	if err != nil {
+		log.Warn("Unable to execute nginx config duration. It will be out of date.", err)
```

Comment:
Same as above, expand the error and let top level take care of logging a problem.